instance/models/server.py

```diff
@@ -321,38 +325,45 @@ def vm_not_yet_requested(self):
         """
         return self.status == Status.Pending
 
-    def update_status(self):
+    def _update_status_from_nova(self, os_server):
         """
-        Refresh the status by querying the openstack server via nova
+        Update the status from the Nova Server object given in os_server.
         """
-        # TODO: Check when server is stopped or terminated
-
-        # First check if it makes sense to update the current status.
-        # This is not the case if we can not interact with the server:
-        if self.status in (Status.BuildFailed, Status.Terminated):
-            return self.status
-        try:
-            os_server = self.os_server
-        except (requests.RequestException, novaclient.exceptions.ClientException):
-            self.logger.debug('Could not reach the OpenStack API')
-            if self.status not in (Status.BuildFailed, Status.Terminated, Status.Pending, Status.Unknown):
-                self._status_to_unknown()
-            return self.status
         self.logger.debug('Updating status from nova (currently %s):\n%s', self.status, to_json(os_server))
-
         if self.status == Status.Unknown:
             if os_server.status in ('INITIALIZED', 'BUILDING'):
                 # OpenStack has multiple API versions; INITIALIZED is current; BUILDING was used in the past
                 self._status_to_building()
-
         if self.status in (Status.Building, Status.Unknown):
             self.logger.debug('OpenStack: loaded="%s" status="%s"', os_server._loaded, os_server.status)
             if os_server._loaded and os_server.status == 'ACTIVE':
                 self._status_to_booting()
-
         if self.status in (Status.Booting, Status.Unknown) and self.public_ip and is_port_open(self.public_ip, 22):
             self._status_to_ready()
 
+    def update_status(self):
+        """
+        Refresh the status by querying the openstack server via nova
+        """
+        # TODO: Check when server is stopped
+
+        # First check if it makes sense to update the current status.
+        # This is not the case if we can not interact with the server:
+        if self.status not in [Status.BuildFailed, Status.Terminated, Status.Pending]:
+            try:
+                os_server = self.os_server
+            except novaclient.exceptions.NotFound:
+                # This exception is raised before the server is created, and after it has been
+                # terminated.  Because of the first "if", we can't get her in Pending state, so the
+                # server must have been terminated.
+                self.logger.debug('Server does not exist anymore: %s', self)
+                self._status_to_terminated()
+            except (requests.RequestException, novaclient.exceptions.ClientException):
+                self.logger.debug('Could not reach the OpenStack API')
+                if self.status != Status.Unknown:
+                    self._status_to_unknown()
```

Comment:
This is the existing behaviour.  The status is updated everytime it's shown in the UI, since the server serializer calls `update_status()`.  And after provisioning the status doesn't do much anyway, apart from being shown in the UI.

I'm not completely sure this behaviour is correct in all cases, but it's what we used so far, and I don't know how to improve it.  I'm also not aware of any major problems it has caused.